pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageCrypto.java

```diff
@@ -0,0 +1,564 @@
+                return false;
+            }
+            dataKeyCipher.init(Cipher.DECRYPT_MODE, privateKey);
+            dataKeyValue = dataKeyCipher.doFinal(encryptedDataKey);
+
+            keyDigest = digest.digest(encryptedDataKey);
+
+        } catch (IllegalBlockSizeException | BadPaddingException | NoSuchAlgorithmException | NoSuchProviderException
+                | NoSuchPaddingException | InvalidKeyException e) {
+            log.error("Failed to decrypt data key {} to decrypt messages {}", keyName, e.getMessage());
+            return false;
+        }
+        dataKey = new SecretKeySpec(dataKeyValue, "AES");
+        dataKeyCache.put(ByteBuffer.wrap(keyDigest), dataKey);
+        return true;
+    }
+
+    private ByteBuf decryptData(SecretKey dataKeySecret, MessageMetadata msgMetadata, ByteBuf payload) {
+
+        // unpack iv and encrypted data
+        ByteString ivString = msgMetadata.getEncryptionParam();
+        ivString.copyTo(iv, 0);
+
+        GCMParameterSpec gcmParams = new GCMParameterSpec(ivLen * 8, iv);
+        ByteBuf targetBuf = null;
+        try {
+            cipher.init(Cipher.DECRYPT_MODE, dataKeySecret, gcmParams);
+
+            ByteBuffer sourceNioBuf = payload.nioBuffer(payload.readerIndex(), payload.readableBytes());
+
+            int maxLength = cipher.getOutputSize(payload.readableBytes());
+            targetBuf = PooledByteBufAllocator.DEFAULT.buffer(maxLength, maxLength);
+            ByteBuffer targetNioBuf = targetBuf.nioBuffer(0, maxLength);
+
+            int decryptedSize = cipher.doFinal(sourceNioBuf, targetNioBuf);
+            targetBuf.writerIndex(decryptedSize);
+
+        } catch (InvalidKeyException | InvalidAlgorithmParameterException | IllegalBlockSizeException
+                | BadPaddingException | ShortBufferException e) {
+            targetBuf.release();
+            targetBuf = null;
+            log.error("Failed to decrypt message {}", e.getMessage());
+        }
+
+        return targetBuf;
+    }
+
+    private ByteBuf decryptDataKeyAndData(MessageMetadata msgMetadata, ByteBuf payload, CryptoKeyReader keyReader) {
+
+        ByteBuf decryptedData = null;
+
+        List<KeyByteValue> encKeys = msgMetadata.getEncryptionKeysList();
+
+        // Go through all keys to retrieve data key from cache
+        for(int i=0; i<encKeys.size(); i++) {
+
+            byte[] msgDataKey = encKeys.get(i).getValue().toByteArray();
+            byte[] keyDigest = digest.digest(msgDataKey);
+            SecretKey storedSecretKey = null;
+            try {
+                storedSecretKey = dataKeyCache.get(ByteBuffer.wrap(keyDigest));
+
+                // Taking a small performance hit here if the hash collides. When it
+                // retruns a different key decryption fails. At this point, we would
+                // call decryptDataKey to refresh the cache and come here again to decrypt.
+                decryptedData = decryptData(storedSecretKey, msgMetadata, payload);
```

Comment:
if `key` is not present then `dataKeyCache` returns `null`. so, instead `decryptData(..)` with null key, can't we just return if `storedSecretKey==null`?

Reply:
I think we may use `Optional` if we are returning null intentionally..

Reply:
null is also returned when decryption fails for other reasons. Also even if it fails with one key, we would attempt to decrypt using the remaining keys.